nvse/nvse/FunctionScripts.h

```diff
@@ -130,38 +130,198 @@ class UserFunctionManager
 {
+	template <bool isSelfOwning>
+	bool PopulateArgs_Templ(ScriptEventList* eventList, FunctionInfo* info,
+		const ArrayElement_Templ<isSelfOwning>* altElemArgs);
+
 public:
 	InternalFunctionCaller(Script* script, TESObjectREFR* callingObj = NULL, TESObjectREFR* container = NULL)
 		: m_callerVersion(UserFunctionManager::kVersion), m_numArgs(0), m_script(script), m_thisObj(callingObj), m_container(container) { }
 
-	virtual ~InternalFunctionCaller() { }
-	virtual UInt8 ReadCallerVersion() {	return m_callerVersion; }
-	virtual Script * ReadScript() {	return m_script; }
-	virtual bool PopulateArgs(ScriptEventList* eventList, FunctionInfo* info);
-	virtual TESObjectREFR* ThisObj() { return m_thisObj; }
-	virtual TESObjectREFR* ContainingObj() { return m_container; }
-	
+	~InternalFunctionCaller() override = default;
+	UInt8 ReadCallerVersion() override { return m_callerVersion; }
+	Script* ReadScript() override { return m_script; }
+
+	bool PopulateArgs(ScriptEventList* eventList, FunctionInfo* info) override;
+
+	TESObjectREFR* ThisObj() override { return m_thisObj; }
+	TESObjectREFR* ContainingObj() override { return m_container; }
+
 	bool SetArgs(UInt8 numArgs, ...);
 	bool vSetArgs(UInt8 numArgs, va_list args);
 	bool SetArgsRaw(UInt8 numArgs, const void* args);
 
+	template <bool isSelfOwning>
+	bool SetArgs(UInt8 numArgs, const ArrayElement_Templ<isSelfOwning> *elemArgs);
+
 protected:
-	enum { kMaxArgs = 10 };	
 
 	UInt8			m_callerVersion;
 	UInt8			m_numArgs;
 	Script			* m_script;
-	void			* m_args[kMaxArgs];
+	void			* m_args[kMaxUdfParams];
+
+	//Points to an array of ArrElems, which retain useful type information.
+	//Used in order to remember if the nth arg was stored as double/int,
+	//since void* loses that info and forces an incorrect conversion if the UDF guessed wrong w/ having a float vs int arg.
+	std::variant<const ArrayElement*, const SelfOwningArrayElement*>
+	m_altElemArgs = static_cast<ArrayElement*>(nullptr);
+
 	TESObjectREFR	* m_thisObj;
 	TESObjectREFR	* m_container;
 
 	virtual bool ValidateParam(UserFunctionParam* param, UInt8 paramIndex) { return param != nullptr; }
-}; 
+};
 
 namespace PluginAPI {
 	bool CallFunctionScript(Script* fnScript, TESObjectREFR* callingObj, TESObjectREFR* container,
 		NVSEArrayVarInterface::Element* result, UInt8 numArgs, ...);
 	bool CallFunctionScriptAlt(Script *fnScript, TESObjectREFR *callingObj, UInt8 numArgs, ...);
 }
 
-#endif
\ No newline at end of file
+
+///// InternalFunctionCaller ///////
+/// (since template funcs must be defined in header) ///
+
+template <bool isSelfOwning>
+bool InternalFunctionCaller::SetArgs(UInt8 numArgs, const ArrayElement_Templ<isSelfOwning>* elemArgs)
+{
+	if (numArgs > kMaxUdfParams)
+		return false;
+
+	m_numArgs = numArgs;
+	m_altElemArgs = elemArgs;
+
+	return true;
+}
+
+template <bool isSelfOwning>
+bool InternalFunctionCaller::PopulateArgs_Templ(ScriptEventList* eventList, FunctionInfo* info,
```

Comment:
This would be better as a separate function from the original one instead of having an if/else in each case statement which checks if altElemArgs exists.

Reply:
Yeah, it felt stupid to have to constantly check each loop if `altElemArgs` was true or not.
Really wanted to avoid having to duplicate the complicated code like
```
var->data = g_StringMap.Add(info->GetScript()->GetModIndex(), str, true);
AddToGarbageCollection(eventList, var, NVSEVarType::kVarType_String);
```
But I suppose it's not so bad.